src/tiered.js

```diff
@@ -23,80 +21,50 @@ class TieredDatastore /* :: <Value> */ {
     this.stores = stores.slice()
   }
 
-  open (callback /* : Callback<void> */) /* : void */ {
-    each(this.stores, (store, cb) => {
-      store.open(cb)
-    }, (err) => {
-      if (err) {
-        return callback(Errors.dbOpenFailedError())
-      }
-      callback()
-    })
+  async open () /* : void */ {
+    try {
+      await (this.stores.map((store) => store.open()))
+    } catch (err) {
+      throw Errors.dbOpenFailedError()
+    }
   }
 
-  put (key /* : Key */, value /* : Value */, callback /* : Callback<void> */) /* : void */ {
-    each(this.stores, (store, cb) => {
-      store.put(key, value, cb)
-    }, (err) => {
-      if (err) {
-        return callback(Errors.dbWriteFailedError())
-      }
-      callback()
-    })
+  async put (key /* : Key */, value /* : Value */) /* : void */ {
+    try {
+      await Promise.all(this.stores.map(store => store.put(key, value)))
+    } catch (err) {
+      throw Errors.dbWriteFailedError()
+    }
   }
 
-  get (key /* : Key */, callback /* : Callback<Value> */) /* : void */ {
-    const storeLength = this.stores.length
-    let done = false
-    let i = 0
-    whilst(() => !done && i < storeLength, cb => {
-      const store = this.stores[i++]
-      store.get(key, (err, res) => {
-        if (err == null) {
-          done = true
-          return cb(null, res)
-        }
-        cb()
-      })
-    }, (err, res) => {
-      if (err || !res) {
-        return callback(Errors.notFoundError())
-      }
-      callback(null, res)
-    })
+  async get (key /* : Key */) /* : Promise<Value> */ {
+    for (const store of this.stores) {
+      try {
+        const res = await store.get(key)
+        if (res) return res
+      } catch (err) {}
+    }
+    throw Errors.notFoundError()
   }
 
-  has (key /* : Key */, callback /* : Callback<bool> */) /* : void */ {
-    const storeLength = this.stores.length
-    let done = false
-    let i = 0
-    whilst(() => !done && i < storeLength, cb => {
-      const store = this.stores[i++]
-      store.has(key, (err, exists) => {
-        if (err == null) {
-          done = true
-          return cb(null, exists)
-        }
-        cb()
-      })
-    }, callback)
+  async has (key /* : Key */) /* : Promise<bool> */ {
```

Comment:
Would it make sense to do this in parallel? It's a little tricky because you want to return true as soon as you get a true response from a store, but it should be possible with a function from [streaming-iterables](https://github.com/bustle/streaming-iterables) (maybe transform()?)